File: swig/csharp/assembly/Logger.cs
// Copyright (c) 2020-2021 Nicholas Corgan
// SPDX-License-Identifier: BSL-1.0

using System;

namespace Pothosware.SoapySDR
{
    /// <summary>
    /// SoapySDR's logging interface.
    ///
    /// Through this, messages can be logged to console or file by registering C# delegates.
    /// The logging API is designed to match System.String.Format().
    /// </summary>
    public class Logger
    {
        /// <summary>
        /// The signature of a function to use for logging.
        /// </summary>
        /// <param name="logLevel">The message's level of importance.</param>
        /// <param name="message">The message text.</param>
        public delegate void LoggerDelegate(LogLevel logLevel, string message);

        private static LoggerDelegate Delegate = null;

        private class CSharpLogHandler : LogHandlerBase
        {
            public CSharpLogHandler() : base()
            {
            }

            public override void Handle(LogLevel logLevel, string message) => Delegate?.Invoke(logLevel, message);
        }

        /// <summary>
        /// The log level threshold. Messages with lower priorities are dropped.
        /// </summary>
        public static LogLevel LogLevel
        {
            get => LogHandlerBase.GetLogLevel();
            set => LogHandlerBase.SetLogLevel(value);
        }

        private static CSharpLogHandler LogHandler = null;

        /// <summary>
        /// Register a custom logging function to be used for all SoapySDR logging calls.
        /// </summary>
        /// <param name="del">A logging function, or null for the default logger (prints to stderr).</param>
        public static void RegisterLogHandler(LoggerDelegate del)
        {
            if(del != null)
            {
                LogHandler = new CSharpLogHandler();
                Delegate = del;
            }
            else
            {
                LogHandler = null;
                Delegate = null;
            }
        }

        /// <summary>
        /// Log a message with a given level and string.
        /// </summary>
        /// <param name="logLevel">The message's priority</param>
        /// <param name="message">The message string</param>
        public static void Log(LogLevel logLevel, string message) => LogHandlerBase.Log(logLevel, message);

        /// <summary>
        /// Log a message with a given level and string, formatted with System.String.Format().
        /// </summary>
        /// <param name="logLevel">The message's priority</param>
        /// <param name="format">The message format</param>
        public static void LogF(LogLevel logLevel, string format, object arg) =>
            Log(logLevel, string.Format(format, arg));

        /// <summary>
        /// Log a message with a given level and string, formatted with System.String.Format().
        /// </summary>
        /// <param name="logLevel">The message's priority</param>
        /// <param name="format">The message format</param>
        public static void LogF(LogLevel logLevel, string format, object[] args) =>
            Log(logLevel, string.Format(format, args));

        /// <summary>
        /// Log a message with a given level and string, formatted with System.String.Format().
        /// </summary>
        /// <param name="logLevel">The message's priority</param>
        /// <param name="format">The message format</param>
        public static void LogF(LogLevel logLevel, IFormatProvider formatProvider, string format, object arg) =>
            Log(logLevel, string.Format(formatProvider, format, arg));

        /// <summary>
        /// Log a message with a given level and string, formatted with System.String.Format().
        /// </summary>
        /// <param name="logLevel">The message's priority</param>
        /// <param name="format">The message format</param>
        public static void LogF(LogLevel logLevel, IFormatProvider formatProvider, string format, object[] args) =>
            Log(logLevel, string.Format(formatProvider, format, args));

        /// <summary>
        /// Log a message with a given level and string, formatted with System.String.Format().
        /// </summary>
        /// <param name="logLevel">The message's priority</param>
        /// <param name="format">The message format</param>
        public static void LogF(LogLevel logLevel, string format, object arg0, object arg1) =>
            Log(logLevel, string.Format(format, arg0, arg1));

        /// <summary>
        /// Log a message with a given level and string, formatted with System.String.Format().
        /// </summary>
        /// <param name="logLevel">The message's priority</param>
        /// <param name="format">The message format</param>
        public static void LogF(LogLevel logLevel, IFormatProvider formatProvider, string format, object arg0, object arg1) =>
            Log(logLevel, string.Format(formatProvider, format, arg0, arg1));

        /// <summary>
        /// Log a message with a given level and string, formatted with System.String.Format().
        /// </summary>
        /// <param name="logLevel">The message's priority</param>
        /// <param name="format">The message format</param>
        public static void LogF(LogLevel logLevel, string format, object arg0, object arg1, object arg2) =>
            Log(logLevel, string.Format(format, arg0, arg1, arg2));

        /// <summary>
        /// Log a message with a given level and string, formatted with System.String.Format().
        /// </summary>
        /// <param name="logLevel">The message's priority</param>
        /// <param name="format">The message format</param>
        public static void LogF(LogLevel logLevel, IFormatProvider formatProvider, string format, object arg0, object arg1, object arg2) =>
            Log(logLevel, string.Format(formatProvider, format, arg0, arg1, arg2));
    }
}

File: luajit/Logger.lua
-- Copyright (c) 2021-2022 Nicholas Corgan
-- SPDX-License-Identifier: BSL-1.0

---
-- Logger API
-- @module SoapySDR.Logger

local ffi = require("ffi")
local lib = require("SoapySDR.Lib")

local Logger =
{
    --- The available priority levels for log messages.
    --
    -- The default log level threshold is INFO.
    -- Log messages with lower priorities are dropped.
    --
    -- The default threshold can be set via the
    -- SOAPY_SDR_LOG_LEVEL environment variable.
    --
    -- Set SOAPY_SDR_LOG_LEVEL to the string value:
    -- "WARNING", "ERROR", "DEBUG", etc...
    -- or set it to the equivalent integer value.
    Level =
    {
        FATAL    = lib.SOAPY_SDR_FATAL,    -- A fatal error. The application will most likely terminate. This is the highest priority.
        CRITICAL = lib.SOAPY_SDR_CRITICAL, -- A critical error. The application might not be able to continue running successfully.
        ERROR    = lib.SOAPY_SDR_ERROR,    -- An error. An operation did not complete successfully, but the application as a whole is not affected.
        WARNING  = lib.SOAPY_SDR_WARNING,  -- A warning. An operation completed with an unexpected result.
        NOTICE   = lib.SOAPY_SDR_NOTICE,   -- A notice, which is an information with just a higher priority.
        INFO     = lib.SOAPY_SDR_INFO,     -- An informational message, usually denoting the successful completion of an operation.
        DEBUG    = lib.SOAPY_SDR_DEBUG,    -- A debugging message.
        TRACE    = lib.SOAPY_SDR_TRACE,    -- A tracing message. This is the lowest priority.
        SSI      = lib.SOAPY_SDR_SSI,      -- Streaming status indicators such as "U" (underflow) and "O" (overflow).
    },

    ---
    -- Send a message to the registered logger.
    -- @tparam Level logLevel a possible logging level
    -- @tparam string message a logger message string
    log = function(logLevel, message)
        lib.SoapySDR_log(logLevel, message)
    end,

    ---
    -- Format a message and send to the registered logger.
    -- The inputs are passed through Lua's string.format().
    -- @tparam Level logLevel a possible logging level
    -- @tparam string format the message format string
    -- @param[opt] ... Optional format arguments
    -- @usage
    -- SoapySDR.Logger.logf(SoapySDR.Logger.Level.INFO, "Logging event %d (%s)", eventNum, eventName)
    logf = function(logLevel, format, ...)
        lib.SoapySDR_log(logLevel, string.format(format, ...))
    end,

    ---
    -- Register a new system log handler.
    -- Platforms should call this to replace the default stdio handler.
    -- @param[opt] handler A function that takes in a log level and message, or nil to restore the default handler 
    -- @usage
    -- function logFunction(logLevel, message)
    --     print(string.format("%s: %s", os.date("%Y%m%d%H%M%S"), message))
    -- end
    --
    -- SoapySDR.Logger.registerHandler(logFunction)
    registerHandler = function(handler)
        -- Transparently handle the Lua string <-> C const char* conversion.
        local internalHandler = function(logLevel, messageFFI)
            return handler(logLevel, ffi.string(messageFFI))
        end

        if handler then lib.SoapySDR_registerLogHandler(internalHandler)
        else lib.SoapySDR_registerLogHandler(nil)
        end
    end,

    ---
    -- Set the log level threshold.
    -- Log messages with lower priority are dropped.
    -- @tparam Level logLevel a possible logging level
    setLevel = function(logLevel)
        lib.SoapySDR_setLogLevel(logLevel)
    end,

    ---
    -- Get the log level threshold
    getLevel = function()
        return tonumber(lib.SoapySDR_getLogLevel())
    end
}

return Logger
